### PR TITLE
Add Status API tests post-overhaul

### PR DESCRIPTION
**Pull request by cvrebert**
This adds the Status API tests promised in https://github.com/jcabi/jcabi-github/pull/1127#discussion_r32581107 and requested in https://github.com/jcabi/jcabi-github/pull/1127#issuecomment-112611522

### CODE REVIEW

**Comment by dmarkov** [comment]
@cvrebert Thanks, I will find someone to review this PR soon

**Comment by dmarkov** [comment]
@carlosmiranda please review, thanks

File: src/test/java/com/jcabi/github/StatusTest.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+     * Name of description property in Status JSON object.
+     */
+    private static final String TARGET_PROP = "target_url";
+
+    /**
+     * Status.Smart can fetch its commit.
+     * @throws IOException If there is an I/O problem.
+     */
+    @Test
+    public void fetchesCommit() throws IOException {
```
Comment (carlosmiranda) [inline src/test/java/com/jcabi/github/StatusTest.java]: @cvrebert Let's declare this method (as well as the other methods in this class) as throwing `Exception`, as prescribed [here](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names).

File: src/main/java/com/jcabi/github/RtStatus.java [hunk 2]
```diff
@@ -42,7 +41,6 @@
  * @version $Id$
  * @since 0.23
  */
-@Immutable
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/RtStatus.java]: @cvrebert Why is this removed? I don't see any changes that impacts this class' immutability.

File: src/test/java/com/jcabi/github/StatusesTest.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    public void convertsToJsonWhenAllPresent() {
+        final String success = "Everything is awesome";
+        MatcherAssert.assertThat(
+            new Statuses.StatusCreate(Status.State.SUCCESS)
+                .withTargetUrl(Optional.of(URL))
+                .withDescription(success)
+                .withContext(Optional.of(CONTEXT))
+                .json().toString(),
+            // @checkstyle LineLength (1 line)
+            Matchers.equalTo("{\"state\":\"success\",\"description\":\"Everything is awesome\",\"context\":\"jcabi/github/test\",\"target_url\":\"http://status.jcabi-github.invalid/42\"}")
```
Comment (carlosmiranda) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @cvrebert This is a little hard to read; let's use `JsonObjectBuilder`.

File: src/test/java/com/jcabi/github/StatusesTest.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+     * Test commit status context string.
+     */
+    private static final String CONTEXT = "jcabi/github/test";
+
+    /**
+     * StatusCreate can convert itself to JSON.
+     */
+    @Test
+    public void convertsToJsonWhenAllPresent() {
+        final String success = "Everything is awesome";
```
Comment (carlosmiranda) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @cvrebert This is a single use variable, we can pass it inline to `withDescription`.

File: src/test/java/com/jcabi/github/StatusesTest.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            Matchers.equalTo("{\"state\":\"success\",\"description\":\"Everything is awesome\",\"context\":\"jcabi/github/test\",\"target_url\":\"http://status.jcabi-github.invalid/42\"}")
+        );
+    }
+
+    /**
+     * StatusCreate can convert itself to JSON when it has no URL.
+     */
+    @Test
+    public void convertsToJsonWhenUrlAbsent() {
+        final String success = "Living the dream!";
```
Comment (carlosmiranda) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @cvrebert Same here, better to pass it directly to `withDescription` as a literal.

File: src/test/java/com/jcabi/github/StatusesTest.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            Matchers.equalTo("{\"state\":\"failure\",\"description\":\"\",\"context\":\"jcabi/github/test\",\"target_url\":\"http://status.jcabi-github.invalid/42\"}")
+        );
+    }
+
+    /**
+     * StatusCreate can convert itself to JSON when it has no context.
+     */
+    @Test
+    public void convertsToJsonWhenContextAbsent() {
+        final String pending = "Kragle is drying...";
```
Comment (carlosmiranda) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @cvrebert Here too, let's pass it directly to `withDescription`.

**Comment by carlosmiranda** [comment]
@cvrebert A few comments above.

File: src/main/java/com/jcabi/github/RtStatus.java (hunk 2 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/RtStatus.java]: @carlosmiranda `@Immutable` currently throws an exception when trying to instantiate this class. Presumably it doesn't think `JsonObject` is immutable.

Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/RtStatus.java]: @cvrebert Hmmm, yes, all right leave it be then.

File: src/test/java/com/jcabi/github/StatusTest.java (hunk 1 above)
Comment (cvrebert) [inline src/test/java/com/jcabi/github/StatusTest.java]: @carlosmiranda Done

File: src/test/java/com/jcabi/github/StatusesTest.java (hunk 4 above)
Comment (cvrebert) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @carlosmiranda This will no longer be single-use after the switch to `JsonObjectBuilder`.

File: src/test/java/com/jcabi/github/StatusesTest.java (hunk 5 above)
Comment (cvrebert) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @carlosmiranda This will no longer be single-use after the switch to `JsonObjectBuilder`.

File: src/test/java/com/jcabi/github/StatusesTest.java (hunk 6 above)
Comment (cvrebert) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @carlosmiranda This will no longer be single-use after the switch to `JsonObjectBuilder`.

File: src/test/java/com/jcabi/github/StatusesTest.java (hunk 4 above)
Comment (carlosmiranda) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @cvrebert makes sense, let's do it.

File: src/test/java/com/jcabi/github/StatusesTest.java (hunk 3 above)
Comment (cvrebert) [inline src/test/java/com/jcabi/github/StatusesTest.java]: @carlosmiranda Done

**Comment by cvrebert** [comment]
@carlosmiranda I've responded to all your comments.

**Comment by carlosmiranda** [comment]
@cvrebert Thanks!

**Comment by carlosmiranda** [comment]
@rultor merge

**Comment by yegor256** [comment]
@cvrebert removing `@Immutable` is a bad idea. We promise that all classes in the library are immutable and annotated this way. See how it's done in other classes, where we encapsulate "JSON-like" strings

**Comment by cvrebert** [comment]
> See how it's done in other classes, where we encapsulate "JSON-like" strings

@yegor256 Actually, FWICT, you just encapsulate *requests* for JSON, which means you re-fetch the JSON over the network every time you need to get an attribute from it. That's extremely inefficient. Can you suggest an **efficient** implementation approach, considering that `@Immutable` doesn't deem `JsonObject` to be immutable?
(I think it's pretty unreasonable to for `@Immutable` to deem all interfaces without `@Immutable` to be mutable; that would require @teamed to control every immutable interface in existence, which is obviously absurd.)

**Comment by cvrebert** [comment]
@yegor256 So, do you have any efficient implementation suggestion? (See previous post)

**Comment by cvrebert** [comment]
@yegor256 So, do you have any efficient implementation suggestion? (See previous post.)

**Comment by yegor256** [comment]
@cvrebert we can encapsulate `String` instead of `JsonObject`...

**Comment by cvrebert** [comment]
@yegor256 Okay, done: 4277a646c2c4bc8ecefc0c9bab62485fd4bc05ed
Still wasteful CPU-wise but at least not network-wise.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by cvrebert** [comment]
@yegor256 Thanks! Would you be able to tag a 0.24 release soon?

**Comment by yegor256** [comment]
@rultor release, tag is `0.24`

**Comment by yegor256** [comment]
@cvrebert there is something wrong with Cobertura... can you take a look?

**Comment by cvrebert** [comment]
@yegor256 Could be an old compiler problem. Try https://github.com/jcabi/jcabi-parent/pull/33

**Comment by cvrebert** [comment]
@yegor256 Okay, please try releasing again now that the other PRs have been merged.

**Comment by yegor256** [comment]
@rultor release, tag is `0.24`

**Comment by yegor256** [comment]
@cvrebert did't help :( try to run `mvn clean install -Pcobertura` locally and you will see the problem

**Comment by cvrebert** [comment]
@yegor256 I will go without the tagged release then. Thanks for trying. Good luck with Cobertura.

**Comment by yegor256** [comment]
@cvrebert ehhh.... it's your code there :) I should have included Cobertura in the "merge" pipeline.. we would fail much earlier than. Are you sure you can't fix this problem?

**Comment by cvrebert** [comment]
@yegor256 I'll give it a shot. Try merging https://github.com/jcabi/jcabi-parent/pull/34 & https://github.com/jcabi/jcabi-parent/pull/35.

**Comment by cvrebert** [comment]
@yegor256 Interesting; `mvn clean install -Pcobertura` doesn't crash for me locally. Perhaps because I'm using Java 8. Could jcabi-github's rultor's Java be upgraded?

**Comment by dmarkov** [comment]
@carlosmiranda I just added **18 mins** to your account, many thanks for your contribution (59896057).. 139 hours and 35 mins spent here. review comments (c=3) added as a bonus. added +18 to your rating, now it is equal to [+3875](http://www.netbout.com/b/35023?open=rating)

**Comment by dmarkov** [comment]
@rultor please deploy

**Comment by yegor256** [comment]
@cvrebert yes, sure, we can use another Docker container, with Java8 inside, but it doesn't seem right to me... looks like there is something "wrong" in the Java code, which we can fix and make more cobertura-friendly. Don't you think?

**Comment by cvrebert** [comment]
@yegor256 Perhaps try un-nesting the class/enum it was complaining about? Anyway, I don't have Java 7 to test with.
